File: supabase/functions/slack-channels/normalize.ts
interface MessageReference {
  message_id: string;
  ref_type: "parent" | "reply";
}

interface NormalizedMessage {
  message_id: string;
  conversation_id: string;
  text: string;
  user: string;
  timestamp: string;
  ref_message_ids: MessageReference[];
}

interface SlackMessage {
  ts: string;
  thread_ts?: string;
  text?: string;
  user?: string;
  [key: string]: unknown;
}

/**
 * Normalizes Slack messages into a flexible structure with ref_message_ids
 *
 * For Slack:
 * - If message has thread_ts and ts !== thread_ts, it's a reply (has parent reference)
 * - If message has thread_ts and ts === thread_ts, it's a parent message (no references yet)
 * - If message has no thread_ts, it's an unthreaded message (no references)
 */
export function normalizeMessages(
  messages: SlackMessage[],
  conversationId: string
): NormalizedMessage[] {
  return messages.map((msg) => {
    const refMessageIds: MessageReference[] = [];

    if (msg.thread_ts && msg.ts !== msg.thread_ts) {
      refMessageIds.push({
        message_id: msg.thread_ts,
        ref_type: "parent",
      });
    }

    return {
      message_id: msg.ts,
      conversation_id: conversationId,
      text: msg.text || "",
      user: msg.user || "",
      timestamp: msg.ts,
      ref_message_ids: refMessageIds,
    };
  });
}

/**
 * Derived property getter: Gets parent message ID from ref_message_ids
 */
export function getParentId(message: NormalizedMessage): string | null {
  const parentRef = message.ref_message_ids.find((ref) => ref.ref_type === "parent");
  return parentRef?.message_id || null;
}

/**
 * Derived property getter: Gets all reply IDs for a given message
 * Note: This requires the full message list to compute
 */
export function getReplyIds(
  message: NormalizedMessage,
  allMessages: NormalizedMessage[]
): string[] {
  return allMessages
    .filter((msg) => getParentId(msg) === message.message_id)
    .map((msg) => msg.message_id);
}

File: supabase/python/slack_channels/normalize.py
from __future__ import annotations

"""Message normalization helpers mirroring the original TypeScript implementation."""

from typing import Literal, Sequence, TypedDict


class MessageReference(TypedDict):
    message_id: str
    ref_type: Literal["parent", "reply"]


class NormalizedMessage(TypedDict):
    message_id: str
    conversation_id: str
    text: str
    user: str
    timestamp: str
    ref_message_ids: list[MessageReference]


class SlackMessage(TypedDict, total=False):
    ts: str
    thread_ts: str
    text: str
    user: str


def normalize_messages(
    messages: Sequence[SlackMessage],
    conversation_id: str,
) -> list[NormalizedMessage]:
    """Normalize Slack history responses into a flat, reference-friendly shape."""

    normalized: list[NormalizedMessage] = []

    for message in messages:
        ts = message.get("ts")
        if not ts:
            # Slack history responses always include ts, but we defensively skip invalid rows.
            continue

        ref_message_ids: list[MessageReference] = []
        thread_ts = message.get("thread_ts")
        if thread_ts and thread_ts != ts:
            ref_message_ids.append({"message_id": thread_ts, "ref_type": "parent"})

        normalized.append(
            {
                "message_id": ts,
                "conversation_id": conversation_id,
                "text": message.get("text", ""),
                "user": message.get("user", ""),
                "timestamp": ts,
                "ref_message_ids": ref_message_ids,
            }
        )

    return normalized


def get_parent_id(message: NormalizedMessage) -> str | None:
    """Return the parent message ID when present."""

    for reference in message["ref_message_ids"]:
        if reference["ref_type"] == "parent":
            return reference["message_id"]
    return None


def get_reply_ids(message: NormalizedMessage, all_messages: Sequence[NormalizedMessage]) -> list[str]:
    """Return the message IDs that reference the provided message as their parent."""

    target_id = message["message_id"]
    return [msg["message_id"] for msg in all_messages if get_parent_id(msg) == target_id]
